File: core/src/main/objc/FoundationProgressIconService.m
/*
 *  Copyright (c) 2012 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

#include <stdio.h>

#import <FoundationProgressIconService.h>
#import <Foundation/Foundation.h>
#import <JavaNativeFoundation/JNFString.h>

JNIEXPORT void JNICALL Java_ch_cyberduck_core_local_FoundationProgressIconService_progress(JNIEnv *env, jobject this, jstring file, jlong current, jlong size)
{
    if(NSClassFromString(@"NSProgress")) {
        id parent = [NSClassFromString(@"NSProgress") performSelector:@selector(currentProgress)];
        if(nil == parent) {
            parent = [NSClassFromString(@"NSProgress") performSelector:@selector(progressWithTotalUnitCount:) withObject:[NSNumber numberWithFloat:size]];
            [parent setPausable:NO];
            [parent setCancellable:NO];
            // Sets the receiver as the current progress object of the current thread.
            [parent becomeCurrentWithPendingUnitCount:size];
            [parent publish];
        }
        else {
            id progress = [NSClassFromString(@"NSProgress") performSelector:@selector(alloc)];
            NSDictionary* info = [NSDictionary dictionaryWithObjectsAndKeys:
                @"NSProgressFileOperationKindDownloading", @"NSProgressFileOperationKindKey",
                [NSURL fileURLWithPath:JNFJavaToNSString(env, file)], @"NSProgressFileURLKey",
                nil];
            [progress performSelector:@selector(initWithParent:userInfo:)
                          withObject:parent
                          withObject:info];
            [progress setKind:@"NSProgressKindFile"];
            [progress setPausable:NO];
            [progress setCancellable:NO];
            [progress setCompletedUnitCount:[[NSNumber numberWithFloat:current] integerValue]];
        }
    }
}

JNIEXPORT void JNICALL Java_ch_cyberduck_core_local_FoundationProgressIconService_cancel(JNIEnv *env, jobject this, jstring file)
{
    if(NSClassFromString(@"NSProgress")) {
        id parent = [NSClassFromString(@"NSProgress") performSelector:@selector(currentProgress)];
        if(nil == parent) {
            return;
        }
        // Balance the most recent previous invocation of becomeCurrentWithPendingUnitCount: on the same thread
        // by restoring the current progress object
        [parent resignCurrent];
    }
}

File: core/dylib/src/main/java/ch/cyberduck/core/local/FoundationProgressIconService.java
package ch.cyberduck.core.local;

/*
 * Copyright (c) 2012 David Kocher. All rights reserved.
 * http://cyberduck.ch/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * dkocher@cyberduck.ch
 */

import ch.cyberduck.binding.foundation.NSArray;
import ch.cyberduck.binding.foundation.NSDictionary;
import ch.cyberduck.binding.foundation.NSProgress;
import ch.cyberduck.binding.foundation.NSString;
import ch.cyberduck.binding.foundation.NSURL;
import ch.cyberduck.core.Local;
import ch.cyberduck.core.transfer.TransferStatus;

import static ch.cyberduck.binding.foundation.NSProgress.*;

public final class FoundationProgressIconService implements IconService {

    @Override
    public boolean set(final Local file, final String image) {
        return false;
    }

    @Override
    public boolean set(final Local file, final TransferStatus status) {
        NSProgress progress = NSProgress.currentProgress();
        if(null == progress) {
            final NSDictionary userInfo = NSDictionary.dictionaryWithObjectsForKeys(
                NSArray.arrayWithObjects(
                    NSString.stringWithString("NSProgressFileOperationKindDownloading"),
                    NSURL.fileURLWithPath(file.getAbsolute())
                ),
                NSArray.arrayWithObjects(NSProgressFileOperationKindKey, NSProgressFileURLKey)
            );
            progress = NSProgress.progressWithParent(null, userInfo);
            progress.setKind(NSProgressKindFile);
            progress.setPausable(false);
            progress.setCancellable(false);
            progress.setCompletedUnitCount(status.getOffset());
            progress.setTotalUnitCount(status.getLength());
            // Sets the receiver as the current progress object of the current thread.
            progress.becomeCurrentWithPendingUnitCount(status.getLength());
            progress.publish();
        }
        else {
            progress.setUserInfoObject_forKey(NSString.stringWithString("NSProgressFileOperationKindDownloading"), NSProgressFileOperationKindKey);
            progress.setUserInfoObject_forKey(NSURL.fileURLWithPath(file.getAbsolute()), NSProgressFileURLKey);
            progress.setCompletedUnitCount(status.getOffset());
            progress.setTotalUnitCount(status.getLength());
        }
        return true;
    }

    @Override
    public boolean remove(final Local file) {
        final NSProgress progress = NSProgress.currentProgress();
        if(null == progress) {
            return false;
        }
        progress.resignCurrent();
        return true;
    }
}
